Random SMSPeer generator.

ricdezen opened the pull request:
## Premise

This is not a class to be used at runtime, it's only useful for testing with consistent quantities of SMSPeers.
Due to special validity criteria of very few countries (Tavana, Norfolk Island, TA (?)) which I am still investigating, the class iterates until an appropriate number is generated.
Any other regions never returned invalid addresses after the first iteration, as long as I could tell.

## Features

An `SMSPeer` generator that can generate numbers for any region, but is limited to 1000 for each, because it simply takes the example number from [libphonenumber](https://github.com/google/libphonenumber) and changes the last three digits, in order not to mess up the national validity of the number.
I don't think there is any better logic for generating phone numbers without implementing the appropriate standards for every single country. If you do, please share.

## Changes

- Added `RandomPeerGenerator` as an interface.
- Added `RandomSMSPeerGenerator` as an implementing class for `SMSPeer`.

## Side Changes

- Added `SMSPeer.getDefaultRegion()` because why in heavens would you want to just blindly set something.

I kept the classes in the main scope to avoid having to import the module with `testImplementation` when using Gradle, if you think moving it to the test scope is worth it, I'll do it.

_Say something nice to who's going to review your code._
[Something nice](https://www.google.com/search?q=cat+pictures&tbm=isch), also thanks for your time.

ricdezen committed:
Merge pull request #1 from Cogno-IDU/master

Pull from Remote

ricdezen committed:
Merge pull request #2 from Cogno-IDU/master

Implemented delivery report (#14)

Commit:
Swapped arguments in MessageParseStrategy

They were reversed compared to the handler and it was bugging me immensely

Commit:
Merge branch 'master' of https://github.com/Cogno-IDU/smslibrary

* 'master' of https://github.com/Cogno-IDU/smslibrary:
  Feature: library working on app closed (#13)
  Created README.md instructions (#16)

Commit:
Merge branch 'master' of https://github.com/Cogno-IDU/smslibrary

# Conflicts:
#	smslibrary/src/main/java/com/eis/smslibrary/SMSMessageHandler.java
#	smslibrary/src/test/java/com/eis/smslibrary/SMSPeerTest.java

Commit:
fix: reverted param order back to original

Commit:
Expand for details:

fix: added Nullable annotations.
feat: added getDefaultRegion static method.

Commit:
Expand for details:

feat: added RandomPeerGenerator as an interface.
feat: added RandomSMSPeerGenerator as a class to generate random SMSPeers.

Commit:
fix: Removed unused constant and reverted mockito version due to powermock bug.

Commit:
fix: made test log more clear.

Commit:
fix: Removed unused fields after Codacy check.

RicDeZen commented on smslibrary/src/main/java/com/eis/random/RandomPeerGenerator.java:
```diff
@@ -0,0 +1,24 @@
+ * @param <P> the type of generated Peer.
+ * @author Riccardo De Zen.
+ */
+public interface RandomPeerGenerator<A, P extends Peer<A>> {
```
Will do, I asked exactly because I couldn't decide, altough this means the library has to be imported as a test library too (I'm 99% sure of this, gotta check)

giovannivelludo replied:
It's not a big deal since it's probably not compiled when the app is built, but I'd have to check how Proguard works to confirm it

RicDeZen commented on smslibrary/src/main/java/com/eis/random/RandomSMSPeerGenerator.java:
```diff
@@ -0,0 +1,113 @@
+
+    /**
+     * Method to generate an invalid address. As a simple invalidity criteria, a random String is
+     * returned.
```
Yeah to make sure the address is invalid I added a header to the generated strings, I should change these specs

RicDeZen commented on smslibrary/src/main/java/com/eis/random/RandomSMSPeerGenerator.java:
```diff
@@ -0,0 +1,113 @@
+                maxDigits = 15,
+                minDigits = 2,
+                maxCharValue = 255;
+        int digits = Math.abs(random.nextInt() % maxDigits) + minDigits;
```
Woops my bad I changed this a couple times and forgot to fix it

RicDeZen commented on smslibrary/src/main/java/com/eis/smslibrary/SMSMessageHandler.java:
```diff
@@ -69,19 +73,21 @@ public String parseData(@NonNull final SMSMessage message) {
          */
         @Override
-        public SMSMessage parseMessage(@NonNull final String channelData, @NonNull final SMSPeer channelPeer) {
+        public SMSMessage parseMessage(@NonNull final String channelData,
```
Yep, happened when pulling from the main repo because I had the reformat option on, did you prefer it like it was before?

giovannivelludo replied:
No, I'd just like to know what shortcut you used

RicDeZen replied:
Just the usual CTRL+ALT+L

giovannivelludo replied:
That's weird, it doesn't automatically wrap text when I use it, maybe I have to change something in my settings

RicDeZen replied:
I may have changed the settings but I've had Android Studio installed for around two years, so I can't really remember, sorry

RicDeZen commented on smslibrary/src/test/java/com/eis/smslibrary/RandomSMSPeerGeneratorTest.java:
```diff
@@ -0,0 +1,116 @@
+@RunWith(Parameterized.class)
+public class RandomSMSPeerGeneratorTest {
+
+    private static final RandomSMSPeerGenerator GENERATOR = new RandomSMSPeerGenerator();
```
Thanks for the reference document, I never really put much thought into this and assumed `static final` -> constant, but now I see, "all Object fields that are never followed by a dot", that is a fair point, method calls could modify the state of an Object and thus its value.
I can't probably track down all of my static final objects named like this, but I will fix at least this and avoid it in the future.

Commit:
Expand for details:

fix: moved classes to test folder
fix: improvements to RandomSMSPeerGenerator specs
fix: moved event validity logging in RandomSMSPeerGeneratorTest to a separate method

giovannivelludo commented on smslibrary/src/test/java/com/eis/smslibrary/random/RandomSMSPeerGenerator.java:
```diff
@@ -0,0 +1,114 @@
+    @NonNull
+    @Override
+    public String generateInvalidAddress() {
+        return getInvalidRandomString();
```
You might as well move the body of `getInvalidRandomString()` in here since it's the only place where it's used.

RicDeZen replied:
I dunno, I feel like leaving it separated in case I need to modify the class later or do something in the address method that doesn't involve the String directly, feels cleaner this way to me

tonin-alessandra reviewed (approved):
Nice, I think this could be really useful for testing!
